### ISSUE TITLE
Contributing your ebuild to Gentoo main repo?

### ISSUE THREAD

**Issue by dr-diem** [dxx-rebirth/dxx-rebirth#502]
Hi,

I see you maintain Gentoo ebuilds in this repo. I recently installed DXX-Rebirth using the ebuild in Gentoo's main repo and found it to be very out of date (0.59.100) - do you mind if I clone your repo so as to track your ebuilds and issue pull requests to the main Gentoo repo? It seems a good idea to keep the core Gentoo repo up to date.

Cheers,

Ian

**Comment by vLKp** [comment]
The ebuilds are under GPL v2, and any use compliant with that license does not need further permission (though I do appreciate being asked, and apprised that someone is working on this).  I'd find it a bit surprising if what you need to do managed to fall even close to a grey area of the GPL. ;)  I hope that the ebuilds are suitable for submission without further change.  If they aren't, please point me to the rejection messages, so that I can address concerns with them.

The ebuilds have fallen behind a bit, so before you submit the snapshot ebuilds (the ones that are not *-9999*), they should be updated to point to a more recent commit.  This is probably just a matter of changing the git commit to which they refer.  You're welcome to try to make that change, or I can make it in my next round of updates.

If you run into any issues getting the ebuilds to work, please report that to this issue tracker, too.  My goal is that, like the source itself, you should be able to get it all working just by following the published instructions.  If you can't, then the instructions need work.

**Comment by dr-diem** [comment]
Thanks!

I'll start with updating the snapshot ebuid to a newer commit (changing the filename appropriately if I step past a release commit) to get a feel for things.

The first major hurdle will be that the current main Gentoo ebuilds reference separate ebuilds for the descent1 and descent2 data, which does feel like a good idea (I leveraged that and that associated USE flags to focus on getting the two executables working individually). If you are in agreement, I will see if I can merge that approach into your (otherwise superior IMO!) ebuild and clone the sub-ebuilds here.

Unrelated, but indeed the installaltion instructions here still make reference to a bug in a point release of gcc 4 - suggest removing that entirely?

**Comment by vLKp** [comment]
> The first major hurdle will be that the current main Gentoo ebuilds reference separate ebuilds for the descent1 and descent2 data

Are you referring to this in the Gentoo ebuild?

https://github.com/gentoo/gentoo/blob/4d96b15775c4144cec3224c8913fac01bf6bd003/games-action/dxx-rebirth/dxx-rebirth-0.59.100.ebuild#L41-L44

If so, that appears to match:

https://github.com/dxx-rebirth/dxx-rebirth/blob/d197ba42c32048a986bfd9e16635528b10bf39e2/contrib/gentoo/games-action/dxx-rebirth/dxx-rebirth-9999.ebuild#L86-L91

The only difference I see is that Gentoo has a USE flag `data` that can be disabled to completely eliminate the dependency on game data, whereas mine will always require you to install some sort of data package for the games you build.  This could be seen as a nuisance if the user wants to manage their game data outside the package manager.  I originally considered it a feature since the games are useless without some sort of data, and people who want to try them out can get the freely redistributable shareware data to satisfy the dependency.

> Unrelated, but indeed the installaltion instructions here still make reference to a bug in a point release of gcc 4 - suggest removing that entirely?

The note was added because a then-supported version of Debian was shipping the bad version of gcc, and someone was impacted by the bug.  Once there are no supported major distributions shipping impacted compiler versions, removing the note would be appropriate.  In the case of Debian, that would mean that they have pushed a fixed version in every release of Debian that still receives updates and that every release of Debian which will not receive the update has been declared end-of-life upstream, so that people using it are aware that they are unsupported.  Until then, I would prefer to leave it in.

**Comment by dr-diem** [comment]
Ah yes, you're right. That's good news then, things will be simpler since I won't need to re-implement the dependency on separate game data ebuilds. I agree with you about the `data` USE flag. If it is easy enough to re-implement it in your ebuild I'll do so, but if it proves incompatible I won't sweat it too much if it is lost. I've a feeling I might be the first person in 4 years to install from the Gentoo ebuild; it only works with gcc 6.5.0 (which I discovered by emerging old versions by trial and error!) whilst the current default is 9.2.0.

Understood regarding the install note about gcc 4; perhaps it is worth adding a reference explaining why it is retained? In my no-prior-experience reading of the installation instructions I took it as a signifier that the instructions hadn't been updated since gcc 4 was the current release ;o)

**Comment by vLKp** [comment]
According to `git blame`, I only added the explicit note about gcc-4.9.2 at the end of 2019.  As I read the [Debian releases page](https://wiki.debian.org/DebianReleases):

- [Debian Jessie](https://wiki.debian.org/DebianJessie) ships with gcc-4.9, reached its first end-of-life 2018-06-17, and will reach its LTS end-of-life around 2020-06-03.  Even now, it is considered the "oldoldstable" distribution.
- [Debian Stretch](https://wiki.debian.org/DebianStretch) ships with gcc-6.3, will reach its first end-of-life sometime this year, and will reach its LTS end-of-life sometime in 2022.  It is currently considered the "oldstable" distribution.
- [Debian Buster](https://wiki.debian.org/DebianBuster) ships with gcc-7 and gcc-8, will reach end-of-life sometime in 2022, and has no LTS end-of-life date reported.  It is currently considered the "stable" distribution.

From that, it's probably safe to drop the note any time.  Although Debian Jessie is not entirely out of support, and thus does not quite match the criterion I described above, it is close to end-of-life.  More importantly, it seems unlikely at this point that anyone would try to package a tip-of-tree Rebirth for Jessie with the end-of-life so close, since Jessie is only receiving major security updates.  Even if they did, the referenced bug may or may not be an issue.  It is a wrong-code bug, which is unfortunate, but I eliminated all known triggers of it.  If it were an obvious build break, renouncing the broken compiler versions would be much easier.

**Comment by vLKp** [comment]
- Raised the minimum version of gcc to gcc-7.  This is the oldest gcc still considered supported by the gcc developers.
- Removed the note about gcc-4.9.2.  Updated various documentation to remove mention of <gcc-7.
- Added `USE=data` support to the Gentoo ebuild.
- Resynchronized the snapshot ebuild with the live ebuild, which gives the snapshot ebuild support for `USE=valgrind`.
- Updated the snapshot ebuild to the latest snapshot, and removed older snapshot ebuilds.

I expect those last few changes should make it easier to get the ebuilds added to the Gentoo tree.

**Comment by dr-diem** [comment]
Thank you - these changes make my life a ton easier!

I will get the Gentoo ebuild updated this week and also add the live one. From there on in I will follow them and update them across as you make releases. I'll follow up here once I've place the pull requests and again when they are accepted.

**Comment by chewi** [comment]
Hello! Gentoo Games lead here. It may surprise you that this is a game I care about a lot. I poured many hours into Decent back in '95. :wink: I did try to bump the package to v0.60 Beta 2 back when it was fresh but it had some non-trivial issues so I waited for the next release... which never really happened. Distros generally prefer actual releases but if you're just going to be doing these occasional snapshots every so often for the foreseeable future then that's fair enough. Ian, I look forward to seeing your work.

**Comment by dr-diem** [comment]
Hi James,

Thanks for the rapid reply and approval to go ahead! Great to hear you're a fellow OG Descent-er :o) I've found the relevant page for how to submit new ebuilds, so I'll (naturally!) do a local test first then file the Bugzilla bug report. If I find similar non-trivial bugs then I'll look to modifying the existing ebuild to indicate that it has a dependency on gcc-6.5.0-r1 (a hoop I had to jump through to get it installed last month!)

Actually, I've also got a question; since I plan to submit both the latest snapshot build and a live build, is it proper form to fill in Bugzilla reports for them individually?

Thanks,

Ian

**Comment by vLKp** [comment]
chewi wrote:
> I did try to bump the package to v0.60 Beta 2 back when it was fresh but it had some non-trivial issues

If you run into packaging problems again, please feel free to open an issue here about it.  I'd like to see all the major Linux distributions able to package and distribute Rebirth with a minimum of downstream patching required.  Patching to make it conform better to distribution policies (preferred filenames and paths, preferred compile options, etc.) is fine if the distribution wants to make changes, but all the basic functionality should work from a fresh checkout without requiring the distribution to patch anything.

> which never really happened.

Sorry about that.  I've never been interested in dealing with the release process, because to me a major part of the release process was offering up prebuilt binaries for Windows users, for Mac users, and possibly for some of the Linux distributions that prefer to ship binaries.  I left that to the other contributors, but they haven't had much time to mess with it in years, so it ended up just not happening.

> but if you're just going to be doing these occasional snapshots every so often for the foreseeable future then that's fair enough.

At this point, I think I am the only person working on Rebirth upstream with any regularity at all.  I can't do Mac builds, and I don't particularly want to deal with providing prebuilt binaries for any of the supported platforms.  If tagging particular commits as releases would be helpful to the Linux distributions, that's easy and I am fine doing that and trying to offer some level of support for the tagged points.

dr-diem wrote:
> If I find similar non-trivial bugs then I'll look to modifying the existing ebuild to indicate that it has a dependency on gcc-6.5.0-r1

I'd prefer that you just ask me to change the code to fix those problems instead of pinning the build on an old compiler, especially since you just talked me into declaring gcc-7 the new minimum. :wink:

**Comment by dr-diem** [comment]
Thanks for the details chewi, and do you know what? As soon as I wrote about modifying the existing ebuild I thought to myself "That's lame - better to tackle the issues!" :)

**Comment by dr-diem** [comment]
Ok, I learnt how to set up a local (Gentoo) repository, cloned your ebuilds there, and have built and run them both with no issues at all using gcc 9.3.0. FWIW my testing has been limited to 'does my last savegame load and can I move, shoot and destroy a robot'; let me know if you have a standard test suite I should go through. The only change I made was to set the KEYWORDS values to ~ in the live (9999) ebuild to mark it as 'testing'; I'll raise a pull request if you like though hardly necessary for adding two ~ chars :o)

Next step is to submit them to Gentoo's master repo and let the gears turn..

**Comment by chewi** [comment]
Sorry if you were waiting on me for anything. I get a constant stream of new bug reports and requests on a daily basis and it's impossible to keep up. I try my best to value contributors such as yourself though. Ultimately we need more official developers who can take responsibility for their commits without having to go through someone else first, even if it's just for a package or two. Unfortunately the bar for entry is a bit high.

**Comment by dr-diem** [comment]
chewi wrote:

> Sorry if you were waiting on me for anything. I get a constant stream of new bug reports and requests on a daily basis and it's impossible to keep up.

No not at all! I was just being distracted by other projects ;o)

> I try my best to value contributors such as yourself though. Ultimately we need more official developers who can take responsibility for their commits without having to go through someone else first, even if it's just for a package or two. Unfortunately the bar for entry is a bit high.

Thanks, and by return I'll try not to add to your burden! Just one question - is [this git procedure](
https://wiki.gentoo.org/wiki/GitHub_Pull_Requests) the correct way forward? That's to say, no need to use the Bugzilla process to notify Gentoo (you?!) of my request to commit new ebuilds?

**Comment by vLKp** [comment]
> let me know if you have a standard test suite I should go through

I do not have a formal play test suite.  I have slowly added some unit tests for various pieces, but they cover only a tiny portion of the total code.

> The only change I made was to set the KEYWORDS values to ~ in the live (9999) ebuild to mark it as 'testing'; I'll raise a pull request if you like though hardly necessary for adding two ~ chars :o)

I thought that the Gentoo convention for live ebuilds was that they have no KEYWORDS, not even `~arch`, because a live ebuild tracks upstream, so the Gentoo maintainers cannot promise any particular level of quality.  If I'm wrong on this and the ebuild should carry keywords, I'm happy to change it.

https://devmanual.gentoo.org/keywording/index.html says:
>> Note: Usually, "live" ebuilds (see Version Control System (VCS) Sources) do not specify a KEYWORDS variable, or assign the empty string to it.

I think that the preferred way for you to test it would be to use `package.accept_keywords` to accept the special keyword `**` for the package, which overrides the keyword check and allows you to **emerge** it despite the lack of keywords.

**Comment by dr-diem** [comment]
Yes, you're right, thanks for pointing that out. I had since added the PROPERTIES+=" live" so I'll now remove the KEYWORDS line entirely.

**Comment by vLKp** [comment]
Good idea.  I added `PROPERTIES=live` to the `-9999` ebuild in the Rebirth tree, to reduce drift between what I store and what you submit.  I probably should change the KEYWORDS line too, but have not done that yet.  I'll wait and see how your submission goes, and review what gets included for other changes to pull back in.

You may also want to look at the `RESTRICT=mirror` in the Rebirth ebuild.  I put that there because otherwise Portage tries to contact Gentoo mirrors to download the source, which doesn't work because Gentoo mirrors don't carry the source since Gentoo isn't carrying this ebuild.  There is no licensing reason Gentoo shouldn't mirror the source, and once the source is on the Gentoo mirrors, that `RESTRICT` line should be removed from the ebuild.

**Comment by dr-diem** [comment]
Yes, another good point - I'd noticed the RESTRICT and will remove it from both ebuilds before committing.

**Comment by dr-diem** [comment]
@vLKp could you correct my partial guesses as to what the following USE flags do:

midi - enable support for MIDI synthesizers
editor - enable support for the level editor?
debug - enable debugging output
sdl2 - ?
tracker - ?

**Comment by vLKp** [comment]
- midi: enforce that SDL-mixer is used and has midi support
- editor: yes
- debug: enable various debugging code paths and sanity checks, not just output.  Some of these checks date back to the retail build, and choke on some custom levels because the game was never coded to sufficiently distrust inputs.
- sdl2: use libSDL2, SDL2-mixer instead of libSDL-1* and SDL-mixer-1*
- tracker: enable support for @Mako88's PHP-based multiplayer game tracker.

**Comment by dr-diem** [comment]
Thanks - pull request now created (https://github.com/gentoo/gentoo/pull/15903) - let's see how it goes.

**Comment by dr-diem** [comment]
Hmmm - the automated QA checks threw a fair number of errors. I will need to alter the layout of the copyright message, we might need the tagged commit to be a tarball instead of a zip, and I fear that it cannot understand the programmatic way you have specified the dependencies :(

**Comment by vLKp** [comment]
- PkgBadlyFormedXml: I do not provide a `metadata.xml`.  It looks like you added one.
- InvalidCopyright: I do not know what this wants changed.
- InvalidLicenseHeader: This looks simply broken: `invalid license header: '#'`
- MissingUnpackerDep: I thought unzip was in `@system`, but it seems not.  However, this may be irrelevant due to the TarballAvailable complaint below.
- NonexistentDeps: `RDEPEND: nonexistent packages: games-action/descent1-freedata, games-action/descent2-freedata`.  These packages exist in my tree.  You may need to get them submitted first in order to resolve this.  You could choose to drop these from the Gentoo version of the ebuild.
- NonsolvableDepsInDev: This might be fallout from the prior point.  If not, I do not know what this wants changed.
- TarballAvailable: I have no objection to using a tarball instead.
- WhitespaceFound: I do not see any trailing whitespace on line 179.  There is a trailing whitespace on line 205 to work around a bash syntax highlighting quirk.  On reviewing your ebuild, I see you removed various lines, including the project's standard phrasing about copyright.  Why did you remove that?  Line 205 in the reference ebuild is the issue.

As a general statement, I discourage modifying copyright-related statements without good cause.

> I fear that it cannot understand the programmatic way you have specified the dependencies :(

What diagnostic led you to this statement?

**Comment by dr-diem** [comment]
Thanks for the detailed feedback!

I do apologize about the copyright statement removal. That was a mistake - it is the historical text statement referencing the code merge that I intend to remove. My next commit will restore the project's standard copyright phrasing, and add a reinforciing statement on the first line (to suit the preferred ebuild style and answer the 'Invalid Copyright' warning).

The diagnostic was actually just lack of coffee. I realised later that I hadn't brought in the dependent packages to my local git repo.

If you could switch your latest ebuild to use tar then I can handle the more minor boilerplate items? The metadata.xml I've already corrected locally.

Thanks again for assisting with this!

Ian

**Comment by dr-diem** [comment]
Okay, things are now in reasonable state. All fatals have been resolved, and the only remaining warnings are complaints that the ebuilds are not Copyright Gentoo Authors (tough!) and missing USE flag metadata for the legacy 0.59 ebuild.

I guess I now wait for the pull request to be merged - @chewi is this on you to progress?

**Comment by vLKp** [comment]
> If you could switch your latest ebuild to use tar then I can handle the more minor boilerplate items?

Done, and snapshot ebuild resynchronized with the -9999 ebuild.

**Comment by dr-diem** [comment]
Hi @vLKp - does your build now automatically include the 'free' maps, or are they still only available separately (as with the existing 0.59.100 ebuild)? I would like to avoid losing features that the existing gentoo ebuild offers you see.

**Comment by vLKp** [comment]
What free maps do you mean?  I see that dxx-rebirth-0.59.100.ebuild has a dependency on `|| ( games-action/descent1-data games-action/descent1-demodata )`.  Mine has the same, albeit generated in a way that may not show up readily on searches:

https://github.com/dxx-rebirth/dxx-rebirth/blob/84fb6d499b7b6efc166f91a3e166a87c4305f39d/contrib/gentoo/games-action/dxx-rebirth/dxx-rebirth-9999.ebuild#L97-L105

I think the answer is that they are only available separately, but my ebuild still prompts the package manager to install them, just as the 0.59.100 ebuild does.

**Comment by dr-diem** [comment]
Thanks, that's how I read it but I'm hyper-cautious of making false assumptions!

The Gentoo maintainer who has been in touch with me insists that I need to clear all the warnings from their Git CI tool in order for him to accept the ebuilds. I need to work out how to honour the USE flags that are unique to the old 0.59.100 ebuild, but the other warning of course is for the copyright statement. He is asking if you would mind transferring your copyright in the three ebuild files to the Gentoo project.

The git history shows  you having committed them originally back in 2017 and that you are the sole contributor, so what do you say? Clearly I wouldn't ask this if it were for the project's 'crown jewels', but if the end result is to give Gentoo users a better DXX-Rebirth experience would you allow it just for the ebuilds? Sorry to be forced to ask :(

**Comment by vLKp** [comment]
> I need to work out how to honour the USE flags that are unique to the old 0.59.100 ebuild

What exactly do you need here?

> He is asking if you would mind transferring your copyright in the three ebuild files to the Gentoo project.

I understand what is meant, but I encourage you to be more careful with using a "would you mind if X" phrasing in serious matters, because, particularly when dealing with non-native English speakers, the negation in the construct can lead to an answer contrary to what the speaker means.  Consider:

> A> Would you mind if I take the last beer?
> B> Yes.  [I don't want you to take the last beer.]
> vs.
> B> No.  [Go ahead, take the last beer.]

I have no intention of executing such a transfer.  The files are intentionally licensed in such a way that no transfer is necessary for Gentoo to freely use these ebuilds.  This is, in my opinion, sufficient for them to be added to the Gentoo Portage tree without a transfer or a special permission grant.  I suggest the maintainer figure out how to get the Git CI tool to understand this.

> Sorry to be forced to ask :(

I understand that you are just the messenger here.

**Comment by dr-diem** [comment]
Thanks for your understanding, and I totally respect your choice.

Personally I agree with you and I think the CI tool is wrong, given that the relevant GLEP ([76](https://www.gentoo.org/glep/glep-0076.html)) does not mandate that Gentoo own copyright in all ebuilds (or indeed any other type of file). Also sorry to have been so very English in forming the question :-D

As for the USE flag warning, that's for me to sort out the metadata.xml to include USE flags for both ebuilds (Gentoo's existing one and your new one), even though some are exclusive to one and not the other; I need to research whether the metadata file format can support this in a form stronger than simply adding text to the comment fields. I will likely modify the d1x and d2x flags to descent1 and descent2 to match the existing ebuild so as to reduce end user confusion, at least until I have the 'power' to drop the 0.59.100 ebuild from the tree ;)

I'll report back when there is further progress..

**Comment by dr-diem** [comment]
I'm about ready to test and commit my tweaks to reference all the USE flags for all three current ebuilds, and the only one I'm not certain about is 'textures'. Am I correct that that it is now a USE flag for the two 'freedata' packages instead?

**Comment by vLKp** [comment]
From looking at your public interactions with the Gentoo project, it is not clear to me what requests were made of you.  Our interactions in this issue are open for the world to see, but the only non-automated interaction I see between you and Gentoo is [one comment from Jonas](https://bugs.gentoo.org/show_bug.cgi?id=724882#c1) telling you to get on IRC and work out the issues there.  Were these interactions logged by the participants' respective IRC clients?  If so, and the participants were aware at the time that they were being logged, I would like to read them.  If not, I suggest that for future discussions with them, you log the conversations (and notify them that you may publish the logs later).

I wonder if some of the problems here may not be a [game of telephone](https://en.wikipedia.org/wiki/Telephone_(game)).  I understand the maintainer's interest in clearing all automated warnings, but based on the unreasonableness of the request[1], and that, as you note, the GLEP does not even mandate such a change, I hope this was a misunderstanding.

[1]: I found the request for a copyright transfer to be not merely unreasonable, but also rather offensive, considering that:
1. I wrote the Rebirth version of these ebuilds without individualized guidance or assistance from anyone associated with the Gentoo project, or even anyone from the Gentoo forums (which is frequented by a broad set of experienced users who volunteer their time for the benefit of the Gentoo project, but are not formally considered part of the Gentoo project).
  I read the public documentation on Gentoo ebuild development, and applied that to what I knew about Rebirth's requirements.  After I was done, one of my revisions attempted to more closely follow their layout so that it would be easier for them to synchronize from me.  I don't recall using their ebuilds for initial design reference, and based on the commit dates, I doubt I could have.  I added my first iteration Sat Apr 8 16:48:19 2017, and chewi added the 0.59.100 ebuilds 2017-04-27 22:18:16, several weeks later.
2. Gentoo maintainers clearly are aware that the Rebirth project is continuing along, because asturm [resolved a Gentoo bug by modifying the 0.59.100 ebuild to apply as a backport one of my fixes for PhysFS](https://gitweb.gentoo.org/repo/gentoo.git/commit/games-action/dxx-rebirth?id=4d96b15775c4144cec3224c8913fac01bf6bd003).  Also, in [that same bug](https://bugs.gentoo.org/show_bug.cgi?id=653014), there is [a comment from chewi](https://bugs.gentoo.org/show_bug.cgi?id=653014#c7) about his unsuccessful attempt to use 0.60.0-beta2, and speculation that a release might be forthcoming.
3. To the best of my recollection, no one from the Gentoo project has ever attempted to coordinate with me, on this or any other aspect of Rebirth.  None of the bug reports in their tracker were forwarded here.  You are the first person I can remember who wanted to synchronize the ebuilds.
4. I have a strong dislike for changes justified on the basis that automated tooling is too simplistic / rigid to permit the work to be done right, so the work is done wrong rather than fixing the tooling to allow the work to be done right.
5. On a purely practical basis, I don't want to mess with performing a copyright transfer.  I have already volunteered countless hours on the Rebirth code itself, and a considerable amount of time helping you help them.  Spending more time to explicitly give away the ebuilds does not make any sense to me.
6. As noted at the start of this issue, to the best of my knowledge, the license that has been on the ebuild for years already provides everything the Gentoo project needs.

Put all that together, and I see no reason why the Gentoo project should receive a copyright transfer of any of this, especially since the ebuilds can equally be useful to derivatives of Gentoo, and so should remain licensed to the benefit of all distributions that can process ebuilds, and indirectly to the benefit of other distributions, where the ebuild can be used as reference material even when the package manager cannot consume it directly.

----

> but if the end result is to give Gentoo users a better DXX-Rebirth experience would you allow it just for the ebuilds?

If you cannot get the maintainer to sort out the bad tooling, your other option would be to discard all your work to date, take the Gentoo 0.59.100 ebuilds, and forward port those to current.  If you don't bother with supporting a live (-9999) ebuild, and instead only do a snapshot, the update process is probably fairly straightforward.
- Rename the ebuild to the new version
- Set an appropriate `SRC_URI`
- Retire any obsolete patches that no longer apply
- Review the DEPEND/RDEPEND blocks

I think that is a rather heavy-handed and backward solution, but it would serve the goal of getting current ebuilds into the main tree.

**Comment by vLKp** [comment]
> I have a question about 'textures'. It is now a USE flag for the two 'freedata' packages, but am I correct that it has to be specified for dxx-rebirth as well in order for its value to fall through to the dependent freedata packages?

That appears to be how the Gentoo project implemented the 0.59.100 ebuilds, but that is because they bundled texture installation into the ebuild that also compiles the engine.  In my implementation, `textures` is only a flag on the data-only ebuilds.  It is not used by the engine ebuild, because it has no effect on the generated code.  You build the same engine whether or not the user has chosen high resolution textures, and the engine will find them - or not - at runtime.  The user can freely add/remove high resolution textures later by adjusting the flags on the freedata package, and an engine rebuild is not required to pick up the results of that addition/removal.  If Portage supported IUSE_RUNTIME, I might have put `textures` on the engine ebuild and in IUSE_RUNTIME, as a way of passing through the preference.  Since Portage does not, I omitted the `textures` flag from the engine ebuild, since it has no effect on the generated code.

In short, no, it does not need to be specified on the dxx-rebirth ebuild.

**Comment by dr-diem** [comment]
Thanks for putting in the effort to stay with me on this, and your considered and detailed responses. I agree with you on all points, and suspect that Jonas has underestimated the complexity of your main ebuild; most do not have programatically generated DEPEND and RDEPEND and scons functions! I did keep a log of the IRC discussion and am more than happy to share it with you, though it's too verbose to paste here. My Git profile has my main email address so please do reach out there and I can send it to you. The salient points are:

- I must clean up the commits that comprise the PR so that it can be more easily understood
- I must find a maintainer for the two freedata packages since they are net-new and I am not a 'recognized' Gentoo contributor (yet!)
- I must clean up all warnings from the Git CI tool

I intend to complete the first two. For the last one I'll resolve all but the copyright warning, for which I will object and point to both [GLEP 76](https://www.gentoo.org/glep/glep-0076.html#id27) and [mgorny's blog post on the topic](https://blogs.gentoo.org/mgorny/2018/09/27/new-copyright-policy-explained/). What stands out for me is that it is a warning and not an error; this distinction exists to allow progress when one understands the warning and knows it not to be a blocker, but that errors need to be corrected before the process can continue.

**Comment by chewi** [comment]
Hello again. I've now caught up on all the correspondence and have done a quick first pass over the submission. I'm replying here as you seem keen to keep the dialogue open.

Pay no mind to Jonas. It's ultimately me you need to keep happy!

I have reread GLEP 76 as I couldn't remember the details. I was vaguely aware that we do have a number of official ebuilds attributed to "Sony Interactive Entertainment Inc." because they make many contributions and require this for their own legal reasons. Between that and the GLEP, I can confirm that there is no strict requirement to put "Gentoo Authors" instead. The warning from repoman is just a warning and although it could be worded better, I think it's mainly there to ensure committers think carefully before entering something else. I will say that "Gentoo Authors" is not intended to strip you of your rights. Rather it seems to be an acknowledgement that keeping track of authors who make minor changes is tedious so you should refer to VCS for details of the precise origins. In fact, your own header does much the same. Perhaps I could make an initial commit with @vLKp as the author to signify the origins in this case but I can understand if you don't want to do that. Ordinarily I might say that this situation could be problematic as we would like the freedom to be able to modify the ebuild as we see fit, not that we cannot, but at some point the copyright header may no longer be appropriate. I can see that you have put a lot of work into these ebuilds though and I can also see that you have a strong intention to continue maintaining them going forward. If we did need to make a change, I am confident that we could work with you in making that happen.

At first glance, the submission is mostly fine. Although I appreciate the symlinked ebuild, I don't think that will fly in the repository so we'll probably just have to duplicate the file. I'd like to see the ebuilds updated to EAPI 7. Regarding the copyright year, I'm not sure whether keeping 2019 is an issue but it looks like we'll be making changes this year anyhow. I'll give them a try myself and review in more detail over the weekend.

**Comment by dr-diem** [comment]
Thanks @chewi !

I agree precisely; this project is active, and I am more than happy to address any issues to bring the ebuilds across, so there's no reason to need to transfer the copyright.

I will duplicate the freedata ebuild to lose the symlink, update all three to EAPI 7 and update the copyright years then report back.

**Comment by vLKp** [comment]
> @dr-diem wrote:
> My Git profile has my main email address so please do reach out there and I can send it to you.

My address is likewise visible in my commits.  However, as we seem to be abandoning the assignment line of thought, the discussion that led to it is not so important now.

> @chewi wrote:
> In fact, your own header does much the same. Perhaps I could make an initial commit with @vLKp as the author to signify the origins in this case but I can understand if you don't want to do that. 

I would suggest that the introductory commit include as part of its log some linkage to the Rebirth ebuild from which the Gentoo ebuild was drawn.  This could be the Rebirth commit ID and repository path from which the ebuild was copied, and some text to let readers readily find the Rebirth repository.  For example:
```
Import Descent Rebirth ebuilds to Gentoo repository

Descent Rebirth repository: https://github.com/dxx-rebirth/dxx-rebirth/
Descent Rebirth commit: d06ad449b8edb8eeea950f0dcc3dfc018df73c0a
Paths:
- contrib/gentoo/games-action/
- contrib/gentoo/games-action/descent1-freedata
- contrib/gentoo/games-action/descent1-freedata/descent1-freedata-1.ebuild
- contrib/gentoo/games-action/descent2-freedata
- contrib/gentoo/games-action/descent2-freedata/descent2-freedata-1.ebuild
- contrib/gentoo/games-action/dxx-rebirth
- contrib/gentoo/games-action/dxx-rebirth/dxx-rebirth-9999.ebuild
- contrib/gentoo/games-action/dxx-rebirth/dxx-rebirth-0.61.0_pre20200523.ebuild
```
Adjust the paths as appropriate for what is actually imported.  The message does not need to look exactly like that.  My concern is that someone who has only the Gentoo Git history can determine where the ebuilds came from, and trace which differences are a result of drift versus intentional deviation.  Subsequent commits written by Gentoo authors would have whatever log messages you think appropriate.  Commits that import more changes from Rebirth could say something like:
```
Cherry-pick Descent Rebirth commit 1234567812345678123456781234567812345678
```

This serves the dual purpose of maintaining the copyright link, and letting subsequent maintainers determine which Rebirth revisions have not yet been brought into Gentoo (and vice versa, if maintainers start contributing their changes back to the Rebirth repository, as I hope they will).

Anything merged to Rebirth would likewise be attributed back to Gentoo, again to ease the task of tracking which changes were not yet moved over.

> @chewi wrote:
> Ordinarily I might say that this situation could be problematic as we would like the freedom to be able to modify the ebuild as we see fit, not that we cannot, but at some point the copyright header may no longer be appropriate.

When it becomes inappropriate, remove it.  I am not fixated on the current header text, and if Gentoo ends up with an ebuild where it no longer makes sense to say that the ebuild is copyright by Rebirth contributors (currently just me, but I'm leaving open the possibility that others will work on this as well), then the header should say something that does make sense.

Similarly, if the current text needs to be reworded to better convey the situation, we can do that.  Perhaps:
```
# Copyright 2020 Gentoo Authors
# Copyright 2017-2020 Rebirth contributors
# Distributed under the terms of the GNU General Public License v2
```
(Years subject to change as needed.)

> @chewi wrote:
> If we did need to make a change, I am confident that we could work with you in making that happen.

I hope so.  Even if you couldn't, the existing license (GNU General Public License v2) should give you the rights you need to maintain and modify a fork without my input or approval, right?

> @chewi wrote:
> Although I appreciate the symlinked ebuild, I don't think that will fly in the repository so we'll probably just have to duplicate the file.

As you wish.  It's convenient for me to maintain it as a link, because I expect the contents will always be the same.  If you need to make structural changes, go ahead.

> @chewi wrote:
> I'd like to see the ebuilds updated to EAPI 7.

Other than just adjusting the assignment itself, is there any porting to be done?

---

Looking at <https://devmanual.gentoo.org/ebuild-writing/eapi/index.html>:

> EAPI 7 Variables
>> PORTDIR and ECLASSDIR are removed
>> DESTTREE and INSDESTTREE are removed

These were not used.
>> D, ED, ROOT, and EROOT modified

`$D` is used, but I think the use of it will handle correctly with and without the trailing `/` because it is prepended to a use of `prefix`, which has a leading `/`.
>> BDEPEND added

Nice improvement, but I think not relevant to Rebirth.  I keep the CBUILD dependencies quite minimal.  I think all that is needed on CBUILD is gcc, Python, and pkg-config (`dev-util/pkgconf` works fine for this).  I think gcc and Python are covered by the rule that `@system` is implied.  I thought pkg-config would be (and for me, `emerge -pve @system` does list `pkgconf` buried in the list), but I am less sure of that.
>> BROOT added

Interesting, but probably not relevant.  I expect all tools to be found on `$PATH` or have their path set explicitly via `scons` arguments.
>> SYSROOT and ESYSROOT added

Not relevant.
>> ENV_UNSET added

Not relevant.

> EAPI 7 Metadata
>> Empty groupings are banned

Not relevant.

> EAPI 7 Profiles
>> package.provided banned

Not relevant.

> EAPI 7 Helpers
>> dohtml banned
>> dolib and libopts banned

Not relevant.  Rebirth does not provide HTML documentation, nor static libraries.
>> has_version and best_version changes

These are not used.
>> Version manipulation and comparison commands
>> New function eqawarn
>> New function dostrip

New features.  They might or might not be useful, but they can be ignored for the sake of porting to EAPI=7 and examined later.
>> die and assert changes
>> nonfatal changes

Nice improvement, but I think not relevant to Rebirth.
>> domo behaviour changed

Rebirth has no localization files.
>> econf changes

Rebirth is not based on autotools, so `econf` is not used.

---

Based on my reading, there is no porting required specifically for EAPI=7.  There may be a need to add a pkg-config package to `BDEPEND`, if the `@system` exemption does not apply.

**Comment by chewi** [comment]
Thanks for the feedback, that all sounds fine.

You're right that pkg-config needs to be in `BDEPEND` but eclasses can also make their own changes around EAPI. In this case, the supported Python versions (for SCons) need to be explicitly specified and the `desktop` eclass now needs to be inherited for `make_desktop_entry` and `doicon`. `eutils` is still needed for `edos2unix`.

I've submitted a review to the pull request. This includes an example of the Python change.

**Comment by vLKp** [comment]
I ran into those Python changes earlier today when I was testing whether the ebuild worked in EAPI=7.  I pushed an update for that a few minutes before I saw your comment.

@dr-diem : you should probably pull in my latest changes:

- Update to EAPI=7
- Add desktop eclass
- Add PYTHON_COMPAT change for `scons-utils`.  I only tried Python 3.7, so that is the only version listed.  I have used Python 3.6 successfully as recently as a few weeks ago, and have not made any changes that I would expect to break Python 3.6 support.  I see that chewi says it works fine in Python 3.8.
- Move KEYWORDS to be more like how I think you will need it (reduces drift / patch maintenance).

I haven't done anything about `BDEPEND` yet.  Would `virtual/pkgconfig` be the most appropriate choice here?

I don't see a use of `edos2unix` in the ebuild I maintain, so there is no `eutils` change yet either.

**Comment by dr-diem** [comment]
Will do @vLKp . There are some additional comments for the freedata packages on the pull request review [here](https://github.com/gentoo/gentoo/pull/15903#pullrequestreview-421530427), just in case you think they too should be fed back.

**Comment by dr-diem** [comment]
Back to working on this today. A couple of questions and comment about your recent changes @vLKp :

- it's not clear to me why you don't have KEYWORDS in your ebuild - surely we need to indicate the supported architectures? I have ~amd64 ~x86; they're set as testing to match the old 0.59 ebuild - I got a CI warning or error if I tried to put them as stable directly.
- since how long has the valgrind keyword been supported? i.e. should I add it to the 20190720 ebuild I am in the process of submitting to Gentoo?
- our favourite the copyright line - given that it is acceptable to have a non-Gentoo copyright, I prefer the older "DXX Rebirth project contributors" since it is clearer than "Rebirth contributors" if someone were to only have the Gentoo ebuild and was trying to work backwards from there to find the original copyright holder. Similarly I favour retaining the detail lines in the freedata packages for similar reasons.

**Comment by chewi** [comment]
We generally don't mark games as stable any more as we don't find the extra work it generates worthwhile. I would expect ~amd64 and ~x86 to be there initially. I could test arm at some point too. Others may want to test further architectures. I'll leave it to @vLKp to decide whether to reflect that here or indeed have any keywords at all.

**Comment by dr-diem** [comment]
Thanks @chewi . @vLKp to add to my comments above, I suspect that this recent change:

```
 DXX_DEPEND_USE_SDL_VERSION_FRAGMENT='
 	media-libs/lib${SDL_version}[joystick?,opengl?,sound,video]
+	media-libs/sdl${SDL_version}-image
 	music? ( media-libs/${SDL_version}-mixer )
 '
```
is not quite correct. According to repoman, ${SDL_version} expands to "sdl" or "sdl2", so prefixing it with "sdl" results in "sdlsdl2" or "sdlsdl" after brace expansion. I think this is correct:

`+	media-libs/${SDL_version}-image`

**Comment by vLKp** [comment]
@dr-diem wrote:

> * it's not clear to me why you don't have KEYWORDS in your ebuild

I do:

https://github.com/dxx-rebirth/dxx-rebirth/blob/e25ecbb10c7786618041c5c0e3af05b16695172a/contrib/gentoo/games-action/dxx-rebirth/dxx-rebirth-9999.ebuild#L18
https://github.com/dxx-rebirth/dxx-rebirth/blob/e25ecbb10c7786618041c5c0e3af05b16695172a/contrib/gentoo/games-action/dxx-rebirth/dxx-rebirth-9999.ebuild#L33

The branch for the live ebuild has `KEYWORDS` explicitly blank, as it is my understanding that is desired for live ebuilds.  The branch for a snapshot ebuild has `KEYWORDS` set to `~amd64 ~x86`.

> * since how long has the valgrind keyword been supported?

According to `git blame`, I first added support for this in 07176ece8b2457e83c6c489be95eeb5c3c9d1d7f, which was posted in September 2014.

---

@chewi wrote:
> Others may want to test further architectures. I'll leave it to @vLKp to decide whether to reflect that here or indeed have any keywords at all.

I can easily test amd64 and x86.  I accept issue reports from other architectures, but have limited ability to debug issues specific to them.  If you or others report how well the game reports on other architectures, I am happy to update `KEYWORDS` to reflect that.

---

@dr-diem wrote:

> I suspect that this recent change:

You are correct.  I verified that `ebuild` would run to completion after the change, but did not run `repoman` on it.  I will correct this.

**Comment by dr-diem** [comment]
Thanks @vLKp - another small issue. The new postint function causes the following error to be thrown:

```
>>> Failed to execute postinst for games-action/dxx-rebirth-0.61.0_pre20190720
>>> Jobs: 2 of 2 complete                           Load avg: 1.53, 0.85, 0.82

 * Messages for package games-action/dxx-rebirth-0.61.0_pre20190720:

 * FAILED postinst: 1
 * ERROR: games-action/dxx-rebirth-0.61.0_pre20190720::carbonrepo failed (postinst phase):
 *   default_pkg_postinst() is not supported in EAPI='7' in phase pkg_postinst
 * 
 * Call stack:
 *            ebuild.sh, line  125:  Called pkg_postinst
 *          environment, line 2242:  Called default
 *   phase-functions.sh, line  855:  Called default_pkg_postinst
 *   phase-functions.sh, line  848:  Called die
```
I resolved it by simply removing the "default" line at the start of the postinst function - I trust this does not break any intended functionality?

**Comment by chewi** [comment]
I don't think the default `pkg_postinst` function does anything so that should be fine but the presence of this function is shadowing `xdg_pkg_postinst`, which is important, so you should call that instead.

**Comment by dr-diem** [comment]
Thanks @chewi - duly added. My local testing is now completed successfully. Barring disaster, by some point this evening I will have completed the commit admin that is still outstanding and consequently have a set of three packages ready for your review.

**Comment by dr-diem** [comment]
Changes all committed; awaiting your feedback @chewi

**Comment by dr-diem** [comment]
Hi again @chewi , as discussed by email, given the interleaved commits for the three new packages, I found it simpler to retract the PR and raise a new one comprising three clean, single-package commits: https://github.com/gentoo/gentoo/pull/16308

**Comment by dr-diem** [comment]
Hi @vLKp - now that the ebuilds are upgraded to EAPI=7, you should alter both to use `xdg_pkg_postinst` instead of `default` in pkg_postinst, per [@chewi's comment](https://github.com/dxx-rebirth/dxx-rebirth/issues/502#issuecomment-644112338) a couple of weeks ago.

**Comment by vLKp** [comment]
Already done as of several days ago.  I've been busy and had not pushed out my recent changes.

**Comment by dr-diem** [comment]
Closing this issue now that a testing and live ebuild have been accepted into the Gentoo main repository.

**Comment by chewi** [comment]
Hello. A couple of things.

@dr-diem, I thought you were going to follow up with the removal of the old ebuild version. It needs to go now due to its use of Python 2, are you good for me to go and do that?

@vLKp, I've just been asked about the copyright header and it turns out there was a reason why repoman was it flagging up. I didn't know that the string needs to be added to this [AUTHORS](https://gitweb.gentoo.org/repo/gentoo.git/tree/metadata/AUTHORS) file to be considered valid. Please could you file a Gentoo bug report about this as it instructs to get this resolved. I would do it but it probably needs to come from you.

**Comment by dr-diem** [comment]
On Mon, 14 Sep 2020, at 16:17, James Le Cuirot wrote:
> 

> Hello. A couple of things.

> @dr-diem <https://github.com/dr-diem>, I thought you were going to follow up with the removal of the old ebuild version. It needs to go now due to its use of Python 2, are you good for me to go and do that?


Yeah, I wanted to leave it a while until we got no bug reports for the new one, and figured it was a 'low-traffic' ebuild so that I should leave it a good 6 months. That said, I am more than happy for you to drop it now as part of the general Python 2 EOL (goodbye and good riddance!). Ironic really - just this weekend I was assaying what Python 2 dependencies I still have.

> @vLKp <https://github.com/vLKp>, I've just been asked about the copyright header and it turns out there was a reason why repoman was it flagging up. I didn't know that the string needs to be added to this AUTHORS <https://gitweb.gentoo.org/repo/gentoo.git/tree/metadata/AUTHORS> file to be considered valid. Please could you file a Gentoo bug report about this as it instructs to get this resolved. I would do it but it probably needs to come from you.


Ah! Glad you got to the bottom of that, because it did seem odd.

Ian

**Comment by chewi** [comment]
@vLKp I've been pinged about this again and if we don't resolve this, it's highly likely that all this hard work will have to be dropped. This should only take a couple of minutes of your time, which is far less than it will take me to recreate these ebuilds from scratch without consulting the existing ones.

**Comment by ulm** [comment]
&lt;Gentoo license team&gt;
Just for my understanding, I suppose that "DXX Rebirth project contributors" is not a legal entity of its own, but simply a wild-card for the copyright holder, in a similar way that "Gentoo Authors" is?

Our [policy](https://www.gentoo.org/glep/glep-0076.html#simplified-attribution) in the main Gentoo repository is that the copyright header *must* say "Gentoo Authors" without any exceptions. The ebuild authors can be listed in our [`metadata/AUTHORS`](https://gitweb.gentoo.org/repo/gentoo.git/tree/metadata/AUTHORS) file if they want (but as you can see in the file, very few of then actually choose to be).
&lt;/Gentoo license team&gt;

**Comment by chewi** [comment]
As stated above, my understanding of the AUTHORS file was slightly wrong. You can add to this file but ebuilds themselves must carry the Gentoo header. @vlkp, I recognise this is not what you wanted but I don't make the rules. I'm just very upset that users will end up losing out if we cannot resolve this.

You said that you didn't believe you had used Gentoo's ebuilds as inspiration for yours but it's been pointed out to me that you most likely did. I'm not at all suggesting you were trying to mislead here as you didn't seem sure and it can be hard to remember. Indeed, I would have realised this sooner if I'd remembered the details myself. I certainly have no desire to be territorial regarding this code, I just hope this discovery can defuse the situation.

The first combined dxx-rebirth ebuild was [submitted to the gamerlay overlay](https://gitweb.gentoo.org/proj/gamerlay.git/commit/?id=1b72eb9b435d23162ac43a176c6202809e419cb9) by @Totktonada in 2016 before you committed yours in 2017. In particular, some of the local USE flag names and variable names in `src_install` match in a way that cannot be coincidental. Granted, gamerlay is not really an official Gentoo repository and @Totktonada has never been a Gentoo developer. However, the ebuild was committed bearing the Gentoo header and when I contacted @Totktonada on 2017-03-06 about bringing his work into the main repository, he was very much in favour. As you saw, I didn't actually get it into the main repository until 2017-04-27, after I'd made some changes.

**Comment by dr-diem** [comment]
Hi @vLKp ,

I'm re-opening this issue in hopes that doing so will bring @chewi's recent posts to your attention. I sincerely hope you'll agree to the proposal - we would so much like to integrate DXX-Rebirth into Gentoo in an ongoing manner!

Please let us know your thoughts,

Ian

**Comment by vLKp** [comment]
@chewi wrote:
> @vLKp I've been pinged about this again and if we don't resolve this, it's highly likely that all this hard work will have to be dropped. This should only take a couple of minutes of your time, which is far less than it will take me to recreate these ebuilds from scratch without consulting the existing ones.

What is there to resolve?  The ebuilds are already under a license that should cover everything you need.  They are in the main tree, and I have not made further changes to them since June.  I suppose I should bump the snapshot ebuild to current, but that's mechanical, and probably not what you mean.

@ulm wrote:
> Just for my understanding, I suppose that "DXX Rebirth project contributors" is not a legal entity of its own, but simply a wild-card for the copyright holder, in a similar way that "Gentoo Authors" is?

That is correct.

> Our policy in the main Gentoo repository is that the copyright header _must_ say "Gentoo Authors" without any exceptions. The ebuild authors can be listed in our metadata/AUTHORS file if they want (but as you can see in the file, very few of them actually choose to be).

That seems like a strange policy.

@chewi wrote:
> As stated above, my understanding of the AUTHORS file was slightly wrong. You can add to this file but ebuilds themselves must carry the Gentoo header. @vLKp, I recognise this is not what you wanted but I don't make the rules. I'm just very upset that users will end up losing out if we cannot resolve this.

What I wanted was not to deal with this strange policy, which has already cost me far more time than it seems likely ever to save me.  I even offered a suggestion back on 2020-05-31 that, I thought, would satisfy the enforcement tool that seems to be driving your request.  As I stated back then, my concern is about letting people who have one git history find the right point in the other git history, so that forward/backward ports are not unnecessarily painful.

To be clear, I'm not the one saying you need to remove the ebuild from the Gentoo tree.  I have no objection to the ebuild in the Gentoo tree, at least as it stands in the commit I see now.  If users get hurt because a Gentoo developer chooses to remove the ebuild, that is on whoever removed it.

> You said that you didn't believe you had used Gentoo's ebuilds as inspiration for yours but it's been pointed out to me that you most likely did. I'm not at all suggesting you were trying to mislead here as you didn't seem sure and it can be hard to remember.

Github is currently hiding my prior comments behind a collapsible banner that it refuses to open, so I cannot go back and see exactly what was written.  I probably would have inspected `emerge -pv` output for the main tree, and matched USE flags to ease the upgrade path.  Was there something more?  Perhaps more importantly, why is this relevant?

Since you want to discuss history, I will note that based on my commit message in e53e969d923956a4019e33993c132001a59f805a, I think I did at that point spend the time to do a line-by-line comparison between the then-current Gentoo ebuild and the one I was maintaining.  The result of that comparison included "match layout with that ebuild where possible, to simplify sharing changes.", as well as a bulleted list of places that were intentionally different (in most cases, because I thought the then-current Gentoo ebuild was incomplete (no support for enabling the editor), incorrect (useless patches), or insufficiently flexible (overzealous blocker relation, mandatory joystick support)).

> The first combined dxx-rebirth ebuild was submitted to the gamerlay overlay by @Totktonada in 2016 before you committed yours in 2017.

I'm not familiar with the gamerlay overlay.  I very much doubt I would have inspected anything in it, or even been aware the ebuild was there.

@dr-diem wrote:
> I'm re-opening this issue in hopes that doing so will bring @chewi's recent posts to your attention.

Good choice.  If I am short on time, which these days is quite often, I visit the issues list without logging in (which by default shows only open issues), look for any recent activity of interest, and log in only if I see something for which a login is required.  Closed issues are invisible to me in this state, and show up only if I log in and check pending notifications.

> I sincerely hope you'll agree to the proposal - we would so much like to integrate DXX-Rebirth into Gentoo in an ongoing manner!

If you are referring to the proposal I previously rejected, where I sign and send in a copyright assignment, I will reiterate my position: I have no interest in making such an assignment.  I see no value to me, or to Rebirth generally, in doing so.  I am frankly rather annoyed that I have now spent more time dealing with pointless policy arguments than I have dealing with actual technical issues.

**Comment by dr-diem** [comment]
Thanks for taking the time to reply @vLKp, and I apologize that these policy wranglings are taking more time than the actual technical work. For what it's worth I too don't understand Gentoo's intransigence with respect to the ebuild copyright header, but we are not able to alter it. As someone who is affiliated neither with the Gentoo nor Rebirth projects, I can contribute my two cents as to the value I see in transferring copyright of the file, in hopes of changing your mind.

The benefit that I see to you and the Rebirth project is that Gentoo remains one of the handful of thriving 'long term' Linux distributions that has a vigorous group of gaming users and, perhaps most importantly, it cares strongly about good quality software, as clearly do you. In the end, having an ebuild in Gentoo's own repo will grow the userbase of Rebirth, which can surely only be a good thing?

Whether spurious or not, Gentoo's policy means that if we cannot transfer copyright in the ebuild, they will drop the ebuild entirely and we'll be back to square one, wherein Gentoo users are forced to either create themselves a local overlay in order to use your ebuild (not trivial) or alternatively install dependencies manually and build Rebirth from source. Should Rebirth be available only to those who know their way around compilation? That seems a great shame to me.

I agree that Gentoo's ebuild policy is strange (@chewi could you point us at an explanation of why it is what it is? Is there a GLEP or discussion thread that led to the current position?), but surely growing the user base and simplifying installation for users of one Linux distro are benefits that are of value to the Rebirth project?

Whatever the historical truth, whatever the ethics of transferring copyright, I think that the community gets a net benefit if, thanks to all your hard work, Descent and Descent 2 can be played by the largest community possible. I appeal to your better nature to reconsider.

**Comment by ulm** [comment]
> Whether spurious or not, Gentoo's policy means that if we cannot transfer copyright in the ebuild, [...]

There seems to be a misunderstanding here. Gentoo is neither asking for signing of a CLA nor for any transfer of copyright. Copyright of all contributions stays with their authors.

> I agree that Gentoo's ebuild policy is strange (@chewi could you point us at an explanation of why it is what it is? Is there a GLEP or discussion thread that led to the current position?), [...]

There were many discussions, but I believe that the following two postings from 2018 summarize the rationale:
https://archives.gentoo.org/gentoo-project/message/9b3b723aa81fb212c7bf0e75efdffe05
https://archives.gentoo.org/gentoo-project/message/ff2533bd5aa32ebf71fcda91ca1e06ec

**Comment by dr-diem** [comment]
Thanks for the clarifications @ulm!

Does this help @vLKp? No transfer of copyright is requested. DXX Rebirth Project Contributors would be listed in the AUTHORS file and, for the purposes of the ebuild, you would be considered to be part of "Gentoo Authors". What do you think (in concert with my other appeal above :) ?

**Comment by vLKp** [comment]
The linked explanation seems to further confuse the issue.  The [second post](https://archives.gentoo.org/gentoo-project/message/ff2533bd5aa32ebf71fcda91ca1e06ec), written by @ulm, closes with a paragraph that seems to directly contradict the need to revisit any of this:

```
The purpose of the proposal explicitly is *not* to stop anybody from
adding an ebuild with a preexisting copyright notice.  However, we should
make it clear that we strongly prefer the simplified form for ebuilds in
the Gentoo repository, entirely for practical reasons.
```

@dr-diem added an ebuild with a non-conforming copyright line, and everything was fine for months.  Then suddenly this became a problem again.  Apparently, "strongly prefer" became "require" sometime after that thread was written.

Regardless, if Gentoo policy requires using a misleading attribution line, go ahead and change it to say "Gentoo Authors" to satisfy whatever mindless tool is complaining about this.  As @ulm noted in that mailing list post, the presence or absence of accurate attribution doesn't mean anything.  (Also, as I noted in my very first comment responding to @dr-diem, I am doubtful someone would find a useful way to use the ebuild such that the conduct was disallowed by the GPL.)

**Comment by dr-diem** [comment]
Thank you @vLKp, your decision is, genuinely, heartily appreciated. I'll make the changes described above and we can move on to the future where I'm able maintain the Gentoo support for Rebirth in the long term.

Thanks all for your input,

Ian
